commands/hugo_test.go

```diff
@@ -0,0 +1,356 @@
+package commands
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/hugo/helpers"
+	"github.com/spf13/hugo/hugofs"
+	"github.com/spf13/hugo/hugolib"
+	jww "github.com/spf13/jwalterweatherman"
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+)
+
+func clearConfigFile() {
+	viper.SetConfigFile("")
+	theme = ""
+	cfgFile = ""
+	destination = ""
+	source = ""
+	baseURL = ""
+}
+
+func TestUserErrorCommandNewUserError(t *testing.T) {
+	defer viper.Reset()
+	expectedUserError := commandError{"test\n", true}
+	actualUserError := newUserError("test")
+	assert.Equal(t, expectedUserError, actualUserError)
+}
+
+func TestSystemErrorCommandNewSystemError(t *testing.T) {
+	defer viper.Reset()
+	expectedSystemError := commandError{"test\n", false}
+	actualSystemError := newSystemError("test")
+	assert.Equal(t, expectedSystemError, actualSystemError)
+}
+
+func TestUserErrorCommandNewFormattedUserError(t *testing.T) {
+	defer viper.Reset()
+	expectedUserError := commandError{"test:%d 12\n", true}
+	actualUserError := newUserError("test:%d", 12)
+	assert.Equal(t, expectedUserError, actualUserError)
+}
+
+func TestSystemErrorCommandNewFormattedSystemError(t *testing.T) {
+	defer viper.Reset()
+	expectedSystemError := commandError{"test:12", false}
+	actualSystemError := newSystemErrorF("test:%d", 12)
+	assert.Equal(t, expectedSystemError, actualSystemError)
+	if actualSystemError != expectedSystemError {
+		t.Errorf("Actual (%v) returned command did not equal expected (%v) command.", actualSystemError, expectedSystemError)
+	}
+}
+
+func TestUserErrorReturnError(t *testing.T) {
+	defer viper.Reset()
+	cError := commandError{"test error", false}
+	expectedError := "test error"
+	actualError := cError.Error()
+	assert.Equal(t, expectedError, actualError)
+}
+
+func TestIsUserError(t *testing.T) {
+	defer viper.Reset()
+	var errors = []struct {
+		err error
+	}{
+		{fmt.Errorf("argument")},
+		{fmt.Errorf("flag")},
+		{fmt.Errorf("shorthand")},
+		{commandError{"test", true}},
+	}
+
+	for _, v := range errors {
+		if !isUserError(v.err) {
+			t.Errorf("Error (%v) should have been flagged as a user error.", v)
+		}
+	}
+}
+
+func TestExecuteAddedCommands(t *testing.T) {
+	defer viper.Reset()
+	defer clearConfigFile()
+	var defaultCommands = []struct {
+		cmd *cobra.Command
+	}{
+		{serverCmd},
+		{versionCmd},
+		{configCmd},
+		{checkCmd},
+		{benchmarkCmd},
+		{convertCmd},
+		{newCmd},
+		{listCmd},
+		{undraftCmd},
+		{importCmd},
+		{genCmd},
+	}
+
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	blogSuffix := strconv.Itoa(r.Intn(math.MaxInt8))
+	basepath := filepath.Join(os.TempDir(), "blog"+blogSuffix)
+	configPath := filepath.Join(basepath, "config.toml")
+	hugofs.Source().MkdirAll(basepath, os.ModePerm|os.ModeDir)
```

Comment:
@moorereason actually, I remember trying that out. But it didn't work. However, I would like to think that I know this stuff a bit better now. :) I'll try again.